engine/common/provider/engine.go

```diff
@@ -212,17 +255,93 @@ func (e *Engine) onEntityRequest(originID flow.Identifier, req *messages.EntityR
 
 	// send back the response
 	res := &messages.EntityResponse{
-		Nonce:     req.Nonce,
+		Nonce:     request.Nonce,
 		EntityIDs: entityIDs,
 		Blobs:     blobs,
 	}
-	err = e.con.Unicast(res, originID)
+	err = e.con.Unicast(res, request.OriginId)
 	if err != nil {
-		return fmt.Errorf("could not send response: %w", err)
+		return engine.NewNetworkTransmissionErrorf("could not send entity response: %w", err)
 	}
 
 	e.metrics.MessageSent(e.channel.String(), metrics.MessageEntityResponse)
 	lg.Debug().Msg("entity response sent")
 
 	return nil
 }
+
+func (e *Engine) processQueuedRequestsShovellerWorker(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	e.log.Debug().Msg("process entity request shoveller worker started")
+
+	for {
+		select {
+		case <-e.requestHandler.GetNotifier():
+			// there is at least a single request in the queue, so we try to process it.
+			e.processAvailableMessages()
+		case <-ctx.Done():
+			// close the internal channel, the workers will drain the channel before exiting
+			close(e.requestChannel)
+			e.log.Trace().Msg("processing entity request worker terminated")
+			return
+		}
+	}
+}
+
+func (e *Engine) processAvailableMessages() {
+	for {
+		msg, ok := e.requestQueue.Get()
```

Comment:
I think we should check if the context was cancelled first and bail early. otherwise this could process forever if there is a consistent stream of new requests

Reply:
https://github.com/onflow/flow-go/pull/3404/commits/ffeb6415857a892e7f1e105c677eae7948439482